itest/lndsigner_test.go

```diff
@@ -0,0 +1,840 @@
+// have a network and optional seed/passphrase, and unixSocket may be used to
+// specify that a UNIX socket should be used to communicate between LND and
+// lndsignerd.
+func (tctx *testContext) addNode(reqPath string,
+	reqData map[string]interface{}, unixSocket bool) string {
+
+	tctx.t.Helper()
+
+	resp, err := tctx.vaultClient.Write(reqPath, reqData)
+	require.NoError(tctx.t, err)
+
+	pubKey, ok := resp.Data["node"].(string)
+	require.True(tctx.t, ok)
+	require.Equal(tctx.t, 66, len(pubKey))
+
+	lnd := &lndHarness{
+		tctx:       tctx,
+		idPubKey:   pubKey,
+		unixSocket: unixSocket,
+	}
+
+	lnd.Start()
+
+	tctx.lndMtx.Lock()
+	tctx.lnds = append(tctx.lnds, lnd)
+	tctx.lndMtx.Unlock()
+
+	return pubKey
+}
+
+// waitForSync ensures that each LND has caught up to the blocks that have been
+// mined on bitcoind.
+func (tctx *testContext) waitForSync() {
+	tctx.t.Helper()
+
+	var (
+		blocks, synced int
+		resp           map[string]interface{}
+	)
+
+	for blocks < 1000 {
+		resp = tctx.bitcoinCli("getblockchaininfo")
+		blocks = int(resp["blocks"].(float64))
+	}
+
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	for _, lnd := range tctx.lnds {
+		synced = 0
+
+		for synced != blocks {
+			time.Sleep(100 * time.Millisecond)
+
+			resp = lnd.Lncli("getinfo")
+			require.Equal(tctx.t, resp["identity_pubkey"].(string),
+				lnd.idPubKey)
+
+			synced = int(resp["block_height"].(float64))
+		}
+	}
+}
+
+// waitForGraphSync ensures that each LND has a synchronized graph.
+func (tctx *testContext) waitForGraphSync() {
+	tctx.t.Helper()
+
+	var (
+		nodes, chans int
+		synced       bool
+		resp         map[string]interface{}
+	)
+
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	for !synced {
+		synced = true
+
+		for _, lnd := range tctx.lnds {
+
+			resp = lnd.Lncli("getnetworkinfo")
+
+			gotNodes := int(resp["num_nodes"].(float64))
+			if gotNodes != nodes {
+				synced = false
+
+				if gotNodes > nodes {
+					nodes = gotNodes
+				}
+			}
+
+			gotChans := int(resp["num_channels"].(float64))
+			if gotChans != chans {
+				synced = false
+
+				if gotChans > chans {
+					chans = gotChans
+				}
+			}
+		}
+
+		time.Sleep(500 * time.Millisecond)
+	}
+}
+
+// testEach runs a function for each LND instance, in parallel.
+func (tctx *testContext) testEach(test func(lnd *lndHarness)) {
+	tctx.t.Helper()
+
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	var wg sync.WaitGroup
+	wg.Add(len(tctx.lnds))
+
+	for _, lnd := range tctx.lnds {
+		innerLnd := lnd
+
+		go func() {
+			defer wg.Done()
+
+			test(innerLnd)
+		}()
+	}
+
+	wg.Wait()
+}
+
+// testEachPair runs a function for each pair of LNDs in parallel, avoiding
+// testing an LND instance with itself.
+func (tctx *testContext) testEachPair(test func(lnd1, lnd2 *lndHarness)) {
+	tctx.t.Helper()
+
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	var wg sync.WaitGroup
+	wg.Add(len(tctx.lnds)*len(tctx.lnds) - len(tctx.lnds))
```

Comment:
Why is it a problem if the wait counter briefly goes to zero? In the end, the counter only matters once you reach `wg.Wait`.

Reply:
You're right, I misunderstood the semantics of the `WaitGroup` where it says "If a WaitGroup is reused to wait for several independent sets of events, new Add calls must happen after all previous Wait calls have returned." That has nothing to do with not calling `Add()` after `Done()`, but only after `Wait()`.

I've fixed this and also used the parallelized `testEach()` in the sweep tests instead of the manual for loop that was there, see https://github.com/bottlepay/lndsigner/compare/d7ba0804a5348e157f205542fa403bf371d91d80..470890621d664c50ed3baafabefc2b0921027a69